beluga_ros/test/test_particle_cloud.cpp

```diff
@@ -48,13 +49,44 @@ TEST(TestParticleCloud, Assign) {
     EXPECT_DOUBLE_EQ(pose.position.z, 0.0);
     EXPECT_DOUBLE_EQ(pose.orientation.x, 0.0);
     EXPECT_DOUBLE_EQ(pose.orientation.y, 0.0);
-    EXPECT_DOUBLE_EQ(pose.orientation.x, 0.0);
-    EXPECT_DOUBLE_EQ(pose.orientation.y, 0.0);
     EXPECT_DOUBLE_EQ(std::abs(pose.orientation.z), std::sin(Constants::pi() / 4.0));
     EXPECT_DOUBLE_EQ(pose.orientation.w, std::cos(Constants::pi() / 4.0));
   }
-  const auto num_particles_in_upper_half_xy_plane =
+  const auto num_particles_in_upper_half_xz_plane =
       ranges::count_if(message.poses, [](const auto& pose) { return pose.position.y > 0.0; });
+  const double upper_half_xz_plane_weight =
+      static_cast<double>(num_particles_in_upper_half_xz_plane) / static_cast<double>(message.poses.size());
+  EXPECT_NEAR(upper_half_xz_plane_weight, 0.5, kSampleSize * 0.01);  // allow 1% deviations
+}
+
+TEST(TestParticleCloud3, Assign) {
+  const auto particles = std::vector{
+      std::make_tuple(
+          Sophus::SE3d{
+              Sophus::SO3d{Eigen::Quaternion<double>{0.8497105, 0.0, 0.3728217, 0.3728217}},
+              Eigen::Vector3d{0.0, 0.0, 1.0}},
+          beluga::Weight(0.5)),
+      std::make_tuple(
+          Sophus::SE3d{
+              Sophus::SO3d{Eigen::Quaternion<double>{0.8497105, 0.0, -0.3728217, 0.3728217}},
+              Eigen::Vector3d{0.0, 0.0, -1.0}},
+          beluga::Weight(0.5)),
+  };
+  constexpr std::size_t kSampleSize = 1000U;
+  auto message = beluga_ros::msg::PoseArray{};
+  beluga_ros::assign_particle_cloud(particles, kSampleSize, message);
+  EXPECT_EQ(message.poses.size(), kSampleSize);
+  for (const auto& pose : message.poses) {
+    EXPECT_DOUBLE_EQ(pose.position.x, 0.0);
+    EXPECT_DOUBLE_EQ(pose.position.y, 0.0);
+    EXPECT_DOUBLE_EQ(std::abs(pose.position.z), 1.0);
+    EXPECT_DOUBLE_EQ(pose.orientation.x, 0.0);
+    EXPECT_NEAR(std::abs(pose.orientation.y), 0.3728217, 0.001);
```

Comment:
@pvela2017 here and elsewhere, consider adding `kTolerance` constants instead of 1e-3 literals.

Reply:
Updated in 442f27f5235d31816a0cdeb67b85ad0671a9c406